text/chapter12.md

```diff
@@ -1,469 +1,285 @@
-# Callback Hell
+# Asynchronous Effects
 
 ## Chapter Goals
 
-In this chapter, we will see how the tools we have seen so far - namely monad transformers and applicative functors - can be put to use to solve real-world problems. In particular, we will see how we can solve the problem of _callback hell_.
+This chapter focuses on the `Aff` monad, which is similar to the `Effect` monad, but represents _asynchronous_ side-effects. We'll demonstrate examples of asynchronously interacting with the filesystem and making HTTP requests. We'll also cover how to manage sequential and parallel execution of asynchronous effects.
 
 ## Project Setup
 
-The source code for this chapter can be compiled and run using `spago run`. It is also necessary to install the `request` module using NPM:
+New PureScript libraries introduced in this chapter are:
 
-```text
-npm install
-```
-
-## The Problem
+- `aff` - defines the `Aff` monad.
+- `node-fs-aff` - asynchronous filesystem operations with `Aff`.
+- `affjax` - HTTP requests with AJAX and `Aff`.
+- `parallel` - parallel execution of `Aff`.
 
-Asynchronous code in JavaScript typically uses _callbacks_ to structure program flow. For example, to read text from a file, the preferred approach is to use the `readFile` function and to pass a callback - a function that will be called when the text is available:
+The `affjax` library requires the `xhr2` NPM module, which you need to install by running:
 
-```javascript
-function readText(onSuccess, onFailure) {
-  var fs = require('fs');
-  fs.readFile('file1.txt', { encoding: 'utf-8' }, function (error, data) {
-    if (error) {
-      onFailure(error.code);
-    } else {
-      onSuccess(data);
-    }   
-  });
-}
+```shell
+$ npm install
 ```
 
-However, if multiple asynchronous operations are involved, this can quickly lead to nested callbacks, which can result in code which is difficult to read:
+## Asynchronous JavaScript
 
-```javascript
-function copyFile(onSuccess, onFailure) {
-  var fs = require('fs');
-  fs.readFile('file1.txt', { encoding: 'utf-8' }, function (error, data1) {
-    if (error) {
-      onFailure(error.code);
-    } else {
-      fs.writeFile('file2.txt', data, { encoding: 'utf-8' }, function (error) {
-        if (error) {
-          onFailure(error.code);
-        } else {
-          onSuccess();
-        }
-      });
-    }   
-  });
-}
-```
+A convenient way to work with asynchronous code in JavaScript is with [`async` and `await`](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Async_await). See [this article on asynchronous JavaScript](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Introducing) for more background information.
```

Comment:
This is just a personal opinion, but I feel like more could be said here about how `Aff` solves some of the problems of async/await / Promises. However, I don't want such an effort to make it sound like we're bashing JS.

Reply:
I'm all for bashing JS (as long as they're accurate bashes and not dated "callback hell" bashes).
What are the issues with async/await though?

Reply:
As far as I could tell, async/await is just the "do notation" of Promises. Referring to a previous comment Nate made, Promises flatten things when they shouldn't, so they aren't entirely monadic. I can't recall where he said that explicitly, but I believe you were a part of that conversation. I think the other issue is that once you're in a Promise context, you can't escape it. I don't know the full implications of why that's an issue, but I recall reading something like that.

Reply:
Flatting promises doesn't seem like a JS flaw, but just something that doesn't translate nicely to PS.
Not sure if the flattening note is going too deep for this book, but if it is mentioned, I think the FFI promises section would be a better spot. https://book.purescript.org/chapter10.html#asynchronous-functions

Also, planning on eventually swapping this ch12 with ch9, so ch10 will have more context.

Reply:
Good point. I think this kind of thing should be documented, but I agree that it could open a can of worms. 
You've already done a lot of work here, so I don't think this should prevent merging this PR.

Reply:
Filed #167 to continue this conversation